### PR TITLE
[EDA] Refactor: REL_Relationships_Con_TDTM - After Insert

### PR DESCRIPTION
**Pull request by m-kinkade**
# Critical Changes

# Changes
### General Updates
To improve readability, modularity, and performance, we've refactored the REL_Relationships_con_TDTM class. The functionality of this class hasn't changed.

# Issues Closed

# New Metadata
### Apex Classes
ContactRelationshipsModel.cls
ContactRelationshipsModel_TEST.cls
ContactRelAutoCreateMappingModel.cls
ContactRelAutoCreateMappingModel_TEST.cls
ContactRelAutoCreateService.cls
ContactRelAutoCreateService_TEST.cls
ContactsRelationshipsModel.cls
ContactsRelationshipsModel_TEST.cls
RelAutoCreateSettingsMapper.cls
RelAutoCreateSettingsMapper_TEST.cls
RelAutoCreateSettingsMapperService.cls
RelAutoCreateSettingsMapperService_TEST.cls

# Deleted Metadata

# Testing Notes
https://salesforce.quip.com/Ii1vARibP5JM#fDdACAHmz0O

### CODE REVIEW

File: src/classes/REL_Relationships_Con_TDTM.cls [hunk 1]
```diff
@@ -41,113 +41,205 @@ public class REL_Relationships_Con_TDTM extends TDTM_Runnable {
...
-            //update type info
-            REL_Relationships_TDTM.addType(listMirrorRels, listRelsForUpdate);
-            dmlWrapper.objectsToUpdate.addAll((list<sObject>)listRelsForUpdate);
+        UTIL_Describe.cacheDescribeSObjectResult(Contact.SObjectType.getDescribe());
+
+        if (triggerAction == TDTM_Runnable.Action.AfterInsert){
+            return this.handleAfterInsert(newList);
         }
 
+        System.debug('Continuing to old code');
```
Comment (xinnanlisf) [inline src/classes/REL_Relationships_Con_TDTM.cls:65]: @m-kinkade saw this while i was working on this class and thought you might want to get rid of this line. :)

File: src/classes/UTIL_CustomSettingsFacade.cls [hunk 2]
```diff
@@ -50,6 +50,56 @@ public without sharing class UTIL_CustomSettingsFacade {
...
+    * @return A list containing the current active Relationship Lookup mappings.
+    ******************************************************************************************************/ 
+    public List<Relationship_Lookup__c> getActiveRelationshipLookups(){
+        if (activeRelationshipLookupList != null){
+            return activeRelationshipLookupList;
+        }
+
+        activeRelationshipLookupList = new List<Relationship_Lookup__c>();
+
+        for (Relationship_Lookup__c relLookup : activeRelationshipLookupList){
```
Comment (xinnanlisf) [inline src/classes/UTIL_CustomSettingsFacade.cls:94]: This doesn't seem right? activeRelationshipLookupList will always be empty at this line? Should we loop through: this.getReciprocalSettings()?

File: src/classes/SRVC_RelAutoCreateSettingsMapper.cls [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        return SRVC_RelAutoCreateSettingsMapper.instance;
+    }
+
+    /***************************************************************************************************************************
+    * @description Retrieves a list of MODL_ConRelationshipAutoCreateMapping representing the Relationship Auto Create
+    * mappings set for the Contact object.
+    * @return A list of ContactRelationshipAutoCreateMapping models.
+    **************************************************************************************************************************/
+    public virtual List<MODL_ConRelationshipAutoCreateMapping> getContactMappingModelsFromSettings(){
+        Map<Schema.SObjectType, List<Relationship_Auto_Create__c>> relAutoCreateListBySObjectType = UTIL_CustomSettingsFacade.getAutoCreateSettingsForContact();
```
Comment (MatthewBlanski) [inline src/classes/SRVC_RelAutoCreateSettingsMapper.cls:68]: Let's add a standard lazy load caching here for a static List<MODL_ConRelationshipAutoCreateMapping>. It wasn't in the original technical spec, but the potential of running multiple schema inspections should be cut off at the tap here.

File: src/classes/SRVC_ContactRelationshipAutoCreate.cls [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+                                                      WHERE Contact__c in :contactsRelationshipsModel.contactIdsSet
+                                                      AND RelatedContact__c in :contactsRelationshipsModel.relatedContactIdSet
+                                                      LIMIT 2000];
+        
+        for (Relationship__c rel : contactRelationships){
+            Relationship__c relToRemove = this.getDuplicateRelationship(rel, contactsRelationshipsModel);
+            if (relToRemove != null){
+                contactsRelationshipsModel.removeRelationshipFromList(relToRemove);
+            } 
+        }
```
Comment (MatthewBlanski) [inline src/classes/SRVC_ContactRelationshipAutoCreate.cls:159]: Given all the guff this is giving us, let's pull down contactsRelationshipsModel.relationshipList first and make a set from it. We can then perform set.remove(relToRemove). We have to cast it back to a list to return it, but this also gives us a side-effect free way to make a new collection.  Were we not dealing with the DMLWrapper downstream, I would simply change this method signature to a Set

**Review by MatthewBlanski — CHANGES_REQUESTED** [review]
There's a lot of good work in here. A few updates need to be made to the logic, but we're in good shape for the most part. There are one or two nasty inefficiencies that could have been introduced based on the types of collections we're using, so let's cut those off instead of introducing a performance-harming measure as part of this refactor.

File: src/classes/REL_Relationships_Con_TEST.cls [hunk 5]
```diff
@@ -0,0 +1,1027 @@
...
+    /*********************************************************************************************************
+    ****************************** RELATIONSHIP AUTO CREATION FUNCTIONAL TESTS *******************************
+    *********************************************************************************************************/
+
+    /*********************************************************************************************************
+    * @description Test to verify that inserting a Contact record without an Auto Creation mapping does not
+    * generate a relationship or reciprocal relationship record.
+    *********************************************************************************************************/
+    @isTest 
+    private static void InsertContactWithoutRelationshipAutoCreate(){
```
Comment (xinnanlisf) [inline src/classes/REL_Relationships_Con_TEST.cls:49]: Saw this while was working on this class. Might want to make the method names lowercase.

File: src/classes/MAPR_RELAutoCreateSettings.cls [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        return MAPR_RELAutoCreateSettings.instance;
+    }
+
+    /***************************************************************************************************************************
+    * @description Retrieves a list of MODL_ConRelationshipAutoCreateMapping representing the Relationship Auto Create
+    * mappings set for the Contact object.
+    * @return A list of ContactRelationshipAutoCreateMapping models.
+    **************************************************************************************************************************/
+    public virtual List<MODL_ConRelationshipAutoCreateMapping> getContactMappingModelsFromSettings(){
+        Map<Schema.SObjectType, List<Relationship_Auto_Create__c>> relAutoCreateListBySObjectType = UTIL_CustomSettingsFacade.getAutoCreateSettingsForContact();
```
Comment (skosgi1) [inline src/classes/MAPR_RELAutoCreateSettings.cls:68]: @MatthewBlanski was this https://github.com/SalesforceFoundation/EDA/pull/1335#discussion_r500080898 done?

**Review by skosgi1 — CHANGES_REQUESTED** [review]
@MatthewBlanski your changes look good, just added a question for https://github.com/SalesforceFoundation/EDA/pull/1335#discussion_r500080898, a few nitpicks on the original PR code and looks like we are missing the test class for MODL_ConRelationshipAutoCreateMapping.

File: src/classes/REL_Relationships_Con_TEST.cls (hunk 5 above)
Comment (MatthewBlanski) [reply src/classes/REL_Relationships_Con_TEST.cls:49]: Done. Thank you.

File: src/package.xml [hunk 7]
```diff
@@ -101,6 +101,13 @@
         <members>FACI_CannotDelete_TEST</members>
         <members>MAPR_PrimaryAffiliations</members>
         <members>MAPR_PrimaryAffiliations_TEST</members>
+        <members>MAPR_RELAutoCreateSettings</members>
+        <members>MAPR_RELAutoCreateSettings_TEST</members>
+        <members>MODL_ConRelationshipAutoCreateMapping</members>
```
[the remark this answers is not in the document]
Comment (MatthewBlanski) [reply src/package.xml:106]: We should.

**Review by MatthewBlanski — DISMISSED** [review]

**Review by skosgi1 — DISMISSED** [review]

File: src/classes/REL_Relationships_Con_TDTM.cls [hunk 8]
```diff
@@ -41,113 +41,204 @@ public class REL_Relationships_Con_TDTM extends TDTM_Runnable {
...
+    private Boolean isInsertRecursion(){
+        return TDTM_ProcessControl.getRecursionFlag(TDTM_ProcessControl.registeredTrigger.REL_Relationships_Con_TDTM);
+    }
     
+    /*******************************************************************************************************
+    * @description Sets the After Insert recursion flag for REL_Relationships_Con_TDTM.
+    *******************************************************************************************************/    
+    @TestVisible
+    private void setInsertRecursion(){
+        TDTM_ProcessControl.setRecursionFlag(TDTM_ProcessControl.registeredTrigger.REL_Relationships_Con_TDTM, true);
```
Comment (xinnanlisf) [inline src/classes/REL_Relationships_Con_TDTM.cls:230]: We have REL_Relationships_Con_TDTM as the insert recursion flag while REL_Relationships_TDTM_After_Update recursion flag. Should we update these flags name in process control class to be more specific?

Comment (MatthewBlanski) [reply src/classes/REL_Relationships_Con_TDTM.cls:230]: We are preserving existing behavior for now.

File: src/classes/SRVC_ContactRelationshipAutoCreate.cls [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+    * @return A MODL_ContactRelationships model representing the Relationships associated with the specified Contact record. 
+    **************************************************************************************************************************/ 
+    @TestVisible
+    protected virtual MODL_ContactRelationships getRelationshipsToCreateForContact(Contact contact, List<MODL_ConRelationshipAutoCreateMapping> contactRelationShipAutoCreateMapModelsList){
+        MODL_ContactRelationships conRelationshipsModel; 
+
+        for (MODL_ConRelationshipAutoCreateMapping conRelAutoCreateMap : contactRelationShipAutoCreateMapModelsList){
+            Object contactFieldValue = contact.get(conRelAutoCreateMap.contactFieldApiName);
+
+            if (contactFieldValue != null){
```
Comment (m-kinkade) [inline src/classes/SRVC_ContactRelationshipAutoCreate.cls:95]: @MatthewBlanski should we throw an error here if the contactFieldValue is not an Id?  The current functionality throws an error for invalid auto create mappings, but it looks like it gets removed in the refactor.  Would it make sense to do something like this:

try{
                    Relationship__c conRelationshipToAdd = new Relationship__c(Contact__c = contact.Id,
                                                                               RelatedContact__c = (Id)contactFieldValue,
                                                                               Type__c = conRelAutoCreateMap.relationshipType,
                                                                               Status__c = 'Current');
    
                    conRelationshipsModel = new MODL_ContactRelationships(contact.Id);
                    conRelationshipsModel.addRelationship(conRelationshipToAdd);
                } catch (Exception e){
                    contact.addError(system.Label.AutoCreateFieldError);
                }

Comment (MatthewBlanski) [reply src/classes/SRVC_ContactRelationshipAutoCreate.cls:95]: Change line 93 to an Id and cast contact.get() to an Id. That'll produce the Id mismatch error. Wrap that line only in the try catch block. Also refer to the Affiliations PR for the exceptions to catch for. Anything other than that we don't want to catch for.

**Review by MatthewBlanski — APPROVED** [review]
This code is functional. There is an outstanding question to ourselves on whether or not we want to rename some of these classes to our new naming convention.

**Review by MatthewBlanski — APPROVED** [review]
These changes look to have taken effect as desired.

**Review by MatthewBlanski — APPROVED** [review]
This is looking good to me. The iterative changes make sense. We would probably review our entire unmanaged unit test strategy

**Comment by spinkelman** [comment]
Passes QA - Merging

**Pull request merged**